fix include outbound ports iptable rules in tproxy mode

ramaraochavali opened the pull request:
As we are discussing https://github.com/istio/istio/issues/27565#issuecomment-765184403 we have an issue with `IncludeOutboundPorts` not working properly in `TPROXY` mode. 

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure


Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ X] Does not have any changes that may affect Istio users.

ramaraochavali committed:
fix include outbound ports in tproxy

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

ramaraochavali commented:
@lambdai see if this makes sense

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | f9789104b74d920a3211580fd307bf5c57b098a2 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30306/lint_istio/1352574869006454784) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | f9789104b74d920a3211580fd307bf5c57b098a2 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30306/lint_istio/1352574869006454784) | `/test lint_istio`
gencheck_istio | f9789104b74d920a3211580fd307bf5c57b098a2 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30306/gencheck_istio/1352574869832732672) | `/test gencheck_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali committed:
fix format

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-security-k8s-tests_istio | 5cec1fe732aa58ae901ed0650c9bad68f1586040 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30306/integ-security-k8s-tests_istio/1352577422133825536) | `/test integ-security-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test integ-security-k8s-tests_istio

rlenglet reviewed (changes requested):
I don't think this makes sense. `ISTIO_OUTPUT` is always a `nat` chain.

ramaraochavali commented:
@rlenglet I see. Thank you.  Based on the ip table output attached here https://github.com/istio/istio/issues/27565#issuecomment-764525722, would you be able to help what is the problem with `IncludeOutboundPorts` in `TPROXY` mode?

lambdai commented:
The current tproxy chain in NAT table is below. see full table in https://github.com/istio/istio/issues/27565#issuecomment-764525722

I don't understand why neither ` -m owner --uid-owner 1337 -j RETURN` nor `-m owner --gid-owner 1337 -j RETURN` hit.
These two rules are supposed to bypass NAT rewrite(target ISTIO_REDIRECT) when Envoy connects to other containers.

@ramaraochavali confirmed that the Envoy is running under uid 1337 in tproxy?

@rlenglet Do you have any clue?

```
-A ISTIO_OUTPUT -o lo -s 127.0.0.6/32 -j RETURN
-A ISTIO_OUTPUT -o lo ! -d 127.0.0.1/32 -m owner --uid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -m owner --uid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -o lo ! -d 127.0.0.1/32 -m owner --gid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -m owner --gid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -d 127.0.0.1/32 -j RETURN
-A ISTIO_OUTPUT -p tcp --dport 50010 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 50020 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 50075 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 50070 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 50090 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 9000 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8019 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8480 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8481 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8485 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 2181 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -d 127.1.2.3/32 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -j RETURN
```

howardjohn commented:
is envoy running as uid 0 for tproxy for root?

On Sat, Jan 23, 2021, 5:28 PM Yuchen Dai <notifications@github.com> wrote:

> The current tproxy chain in NAT table is below. see full table in #27565
> (comment)
> <https://github.com/istio/istio/issues/27565#issuecomment-764525722>
>
> I don't understand why neither -m owner --uid-owner 1337 -j RETURN nor -m
> owner --gid-owner 1337 -j RETURN hit.
> These two rules are supposed to bypass NAT rewrite(target ISTIO_REDIRECT)
> when Envoy connects to other containers.
>
> @ramaraochavali <https://github.com/ramaraochavali> confirmed that the
> Envoy is running under uid 1337 in tproxy?
>
> @rlenglet <https://github.com/rlenglet> Do you have any clue?
>
> -A ISTIO_OUTPUT -o lo -s 127.0.0.6/32 -j RETURN
> -A ISTIO_OUTPUT -o lo ! -d 127.0.0.1/32 -m owner --uid-owner 1337 -j ISTIO_IN_REDIRECT
> -A ISTIO_OUTPUT -m owner --uid-owner 1337 -j RETURN
> -A ISTIO_OUTPUT -o lo ! -d 127.0.0.1/32 -m owner --gid-owner 1337 -j ISTIO_IN_REDIRECT
> -A ISTIO_OUTPUT -m owner --gid-owner 1337 -j RETURN
> -A ISTIO_OUTPUT -d 127.0.0.1/32 -j RETURN
> -A ISTIO_OUTPUT -p tcp --dport 50010 -j ISTIO_REDIRECT
> -A ISTIO_OUTPUT -p tcp --dport 50020 -j ISTIO_REDIRECT
> -A ISTIO_OUTPUT -p tcp --dport 50075 -j ISTIO_REDIRECT
> -A ISTIO_OUTPUT -p tcp --dport 50070 -j ISTIO_REDIRECT
> -A ISTIO_OUTPUT -p tcp --dport 50090 -j ISTIO_REDIRECT
> -A ISTIO_OUTPUT -p tcp --dport 9000 -j ISTIO_REDIRECT
> -A ISTIO_OUTPUT -p tcp --dport 8019 -j ISTIO_REDIRECT
> -A ISTIO_OUTPUT -p tcp --dport 8480 -j ISTIO_REDIRECT
> -A ISTIO_OUTPUT -p tcp --dport 8481 -j ISTIO_REDIRECT
> -A ISTIO_OUTPUT -p tcp --dport 8485 -j ISTIO_REDIRECT
> -A ISTIO_OUTPUT -p tcp --dport 2181 -j ISTIO_REDIRECT
> -A ISTIO_OUTPUT -d 127.1.2.3/32 -j ISTIO_REDIRECT
> -A ISTIO_OUTPUT -j RETURN
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/30306#issuecomment-766272065>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEYGXNNHGT6YCO5GAE2N3LS3NZTDANCNFSM4WOKLZSA>
> .
>

lambdai commented:
@howardjohn Theoretically the requirement is CAP_NET_ADMIN. Uid 0 is overkill.

howardjohn commented:
I recall the injection template may set it to 0 though - worth checking. I
may remember wrong

On Sat, Jan 23, 2021, 5:33 PM Yuchen Dai <notifications@github.com> wrote:

> @howardjohn <https://github.com/howardjohn> Theoretically the requirement
> is CAP_NET_ADMIN. Uid 0 is overkill.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/30306#issuecomment-766272537>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEYGXLX4QR6LIHJQWJQFHTS3N2G3ANCNFSM4WOKLZSA>
> .
>

lambdai commented:
> I recall the injection template may set it to 0 though - worth checking. I may remember wrong
> […](#)

Then I guess the `PROXY_UID=1337, PROXY_GID=1337` alone doesn't imply envoy running uid 1337.

Maybe the injection deserves the change, not the iptables. 
CC @ramaraochavali

ramaraochavali committed:
fix injection template

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

ramaraochavali commented:
@lambdai Thank you so much for looking in to this. Yeah, I also figured out that it is not returning properly but could not think uid is problem - I was tripped off by the iptable output. Looked at the injection template and pod.yaml - Yes, in TPROXY mode it is running as 0 as @howardjohn says. 

I reverted the iptable change and fixed the template. 

@lambdai @howardjohn @rlenglet PTAL - Thanks for helping out on this.

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_istio | f86f2e4758631dc4d27767049ed7435d06afc449 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30306/gencheck_istio/1353220286928916480) | `/test gencheck_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

lambdai commented on manifests/charts/istio-control/istio-discovery/files/injection-template.yaml:
```diff
@@ -343,8 +343,14 @@ spec:
       runAsGroup: 1337
       fsGroup: 1337
       {{ if or (eq (annotation .ObjectMeta `sidecar.istio.io/interceptionMode` .ProxyConfig.InterceptionMode) `TPROXY`) (eq (annotation .ObjectMeta `sidecar.istio.io/capNetBindService` .Values.global.proxy.capNetBindService) `true`) -}}
+      {{ if eq (annotation .ObjectMeta `sidecar.istio.io/interceptionMode` .ProxyConfig.InterceptionMode) `TPROXY` -}}
+      runAsNonRoot: false
+      runAsUser: 1337
+      {{- end }}
+      {{ if eq (annotation .ObjectMeta `sidecar.istio.io/capNetBindService` .Values.global.proxy.capNetBindService) `true` -}}
       runAsNonRoot: false
       runAsUser: 0
+      {{- end }}
```
These 2 `if` are not exclusive. IT means runAsNonRoot and runAsUser could appear twice.

Forgive my lack of knowledge on template language: what is the corresponding `if` of the `else` on line 354?

lambdai replied:
Ok, the else matches the `if` w/o `end`

ramaraochavali replied:
Ah. Will fix.

lambdai commented on pkg/kube/inject/testdata/inject/hello-tproxy.yaml.injected:
```diff
@@ -137,7 +137,7 @@ spec:
           readOnlyRootFilesystem: true
           runAsGroup: 1337
```
See here: gid 1337 should have the same effect as uid 1337.

lambdai replied:
This is not the common output of tproxy setting but a special arg in the test case

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_istio | f86f2e4758631dc4d27767049ed7435d06afc449 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30306/gencheck_istio/1353220286928916480) | `/test gencheck_istio`
integ-pilot-k8s-tests_istio | f86f2e4758631dc4d27767049ed7435d06afc449 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30306/integ-pilot-k8s-tests_istio/1353220281954471936) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali committed:
make gen

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

lambdai commented on manifests/charts/istiod-remote/files/injection-template.yaml:
```diff
@@ -343,8 +343,14 @@ spec:
       runAsGroup: 1337
       fsGroup: 1337
       {{ if or (eq (annotation .ObjectMeta `sidecar.istio.io/interceptionMode` .ProxyConfig.InterceptionMode) `TPROXY`) (eq (annotation .ObjectMeta `sidecar.istio.io/capNetBindService` .Values.global.proxy.capNetBindService) `true`) -}}
+      {{ if eq (annotation .ObjectMeta `sidecar.istio.io/interceptionMode` .ProxyConfig.InterceptionMode) `TPROXY` -}}
+      runAsNonRoot: false
+      runAsUser: 1337
+      {{- end }}
+      {{ if eq (annotation .ObjectMeta `sidecar.istio.io/capNetBindService` .Values.global.proxy.capNetBindService) `true` -}}
       runAsNonRoot: false
       runAsUser: 0
+      {{- end }}
```
ditto

lambdai reviewed (changes requested):
Great! We are on the same page. The uid/gid is the culprit.

I am looking at the hello-tproxy.yaml. The golden test has the `runAsGroup: 1337` that is supposed to match iptables rules.

I am wondering if `runAsGroup:1337` doesn't work as we expected. Maybe `securityContext` as whole, or docker image managing linux [user namespaces](https://man7.org/linux/man-pages/man7/user_namespaces.7.html) was broken.

Anyway, our ultimate goal is to run Envoy 
1. under either uid 1337 or gid 1337, or both.
2. Envoy process must have CAP_NET_ADMIN capability if TPROXY is enabled.

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | b6c4e5396ffbc7b79a182b1c673193f1ef25bdac | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30306/integ-pilot-k8s-tests_istio/1353229646853836800) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | b6c4e5396ffbc7b79a182b1c673193f1ef25bdac | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30306/integ-pilot-k8s-tests_istio/1353229646853836800) | `/test integ-pilot-k8s-tests_istio`
integ-pilot-multicluster-tests_istio | b6c4e5396ffbc7b79a182b1c673193f1ef25bdac | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30306/integ-pilot-multicluster-tests_istio/1353229647633977344) | `/test integ-pilot-multicluster-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/retest

ramaraochavali commented:
@lambdai So we patch the security context in the operator with the following
```
securityContext:
  allowPrivilegeEscalation: false
  capabilities:
    drop:
    - ALL
  privileged: false
  readOnlyRootFilesystem: false
  runAsUser: 1337
```
And the SecurityContext for `TPROXY` pod is
```
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        add:
        - NET_ADMIN
        drop:
        - ALL
      privileged: false
      readOnlyRootFilesystem: true
      runAsNonRoot: false
      runAsUser: 0
```  
and for `REDIRECT` is 

```
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
      privileged: false
      readOnlyRootFilesystem: true
      runAsNonRoot: true
      runAsUser: 1337
```

So I guess the operator patching is ignoring `runAsGroup` as it is not there in the patch?

So I am wondering the following rule is helping in `REDIRECT` mode 
```
-A ISTIO_OUTPUT -m owner --uid-owner 1337 -j RETURN 
```
But since runAsUser is 0 for TPROXY - it is failing.

So this PR is setting UID to 1337 which works for us.

WDYT?

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-multicluster-tests_istio | b6c4e5396ffbc7b79a182b1c673193f1ef25bdac | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30306/integ-pilot-multicluster-tests_istio/1353229647633977344) | `/test integ-pilot-multicluster-tests_istio`
integ-pilot-k8s-tests_istio | b6c4e5396ffbc7b79a182b1c673193f1ef25bdac | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30306/integ-pilot-k8s-tests_istio/1353256520061554688) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | b6c4e5396ffbc7b79a182b1c673193f1ef25bdac | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30306/integ-pilot-k8s-tests_istio/1353256520061554688) | `/test integ-pilot-k8s-tests_istio`
integ-pilot-multicluster-tests_istio | b6c4e5396ffbc7b79a182b1c673193f1ef25bdac | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30306/integ-pilot-multicluster-tests_istio/1353256520103497728) | `/test integ-pilot-multicluster-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali committed:
fix else if

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-multicluster-tests_istio | b6c4e5396ffbc7b79a182b1c673193f1ef25bdac | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30306/integ-pilot-multicluster-tests_istio/1353256520103497728) | `/test integ-pilot-multicluster-tests_istio`
integ-pilot-k8s-tests_istio | de3385c6bca752ed12b1a65fd9f54690e05a1dc6 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30306/integ-pilot-k8s-tests_istio/1353300056177905664) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | de3385c6bca752ed12b1a65fd9f54690e05a1dc6 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30306/integ-pilot-k8s-tests_istio/1353300056177905664) | `/test integ-pilot-k8s-tests_istio`
integ-pilot-multicluster-tests_istio | de3385c6bca752ed12b1a65fd9f54690e05a1dc6 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/30306/integ-pilot-multicluster-tests_istio/1353300056958046208) | `/test integ-pilot-multicluster-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

rlenglet commented:
When did `runAsGroup` ever get removed from the `securityContext`? 🤔

rlenglet reviewed (changes requested):
Let's first understand why `runAsGroup` is not set in the `securityContext` of the container. It's specified in the injection template, but disappeared in the container's context. I don't understand how that's possible.

ramaraochavali commented:
@rlenglet Agreed.  I am still looking in to why `runAsGroup` is not coming even when the injection template has it. One possible explanation is `runAsGroup` on container can be disabled using feature gate https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/. I am checking with our kubernetes team if that is the case in our environment.

Do you see any issues setting user to `1337` for `TPROXY`?

howardjohn commented:
If UID != GID there are some edge cases that may not work btw. I *think* TestTraffic/selfcall will fail

rlenglet commented:
> Do you see any issues setting user to 1337 for TPROXY?

Yes. IIRC, that's because Docker doesn't set ambient capabilities. So when it switches the container's UID from 0 to 1337, it loses all capabilities. Adding capabilities only work if UID is 0.

lambdai commented:
> If UID != GID there are some edge cases that may not work btw. I _think_ TestTraffic/selfcall will fail

That's a good point. The goals of (uid != 1337, gid=1337) are 
1. Don't intercept uid=0 since we assume this uid is shared among Envoy and other processes.
2. Capture on gid=1337 which is Envoy(or whichever don't want to be captured)

So the Envoy runAsUid and captured uid must be independent, with the default case that 
runAsUid = 1337, captured uid = {1337}. Note that captured uid is a collection.

lambdai commented:
> Yes. IIRC, that's because Docker doesn't set ambient capabilities. So when it switches the container's UID from 0 to 1337, it loses all capabilities. Adding capabilities only works if the UID is 0.

Is this the capability uniquely lost in docker-shim CRI?  Other CRI may support

ramaraochavali commented:
Ok. Another finding to test whether gid is the actual issue, I passed `-g 0,1337` to iptables to let it skip gid `0` also. But the problem exists - It makes me wonder it is not gid but some thing else is there

Here is the proxy id `uid=0(root) gid=0(root) groups=0(root),1337(istio-proxy)`

and the ip table rules - now it added a rule to skip if gid is 0 - but still it did not skip

```
Environment:
------------
ENVOY_PORT=
INBOUND_CAPTURE_PORT=
ISTIO_INBOUND_INTERCEPTION_MODE=
ISTIO_INBOUND_TPROXY_MARK=
ISTIO_INBOUND_TPROXY_ROUTE_TABLE=
ISTIO_INBOUND_PORTS=
ISTIO_OUTBOUND_PORTS=
ISTIO_LOCAL_EXCLUDE_PORTS=
ISTIO_SERVICE_CIDR=
ISTIO_SERVICE_EXCLUDE_CIDR=

Variables:
----------
PROXY_PORT=15001
PROXY_INBOUND_CAPTURE_PORT=15006
PROXY_TUNNEL_PORT=15008
PROXY_UID=1337
PROXY_GID=0,1337
INBOUND_INTERCEPTION_MODE=TPROXY
INBOUND_TPROXY_MARK=1337
INBOUND_TPROXY_ROUTE_TABLE=133
INBOUND_PORTS_INCLUDE=*
INBOUND_PORTS_EXCLUDE=15090,15021,8181,15020
OUTBOUND_IP_RANGES_INCLUDE=127.1.2.3/32
OUTBOUND_IP_RANGES_EXCLUDE=
OUTBOUND_PORTS_INCLUDE=50010,50020,50075,50070,50090,9000,8019,8480,8481,8485,2181,8030,8031,8032,8033,8088,8040,8041,8042,10020,19888,33911,8020,8082
OUTBOUND_PORTS_EXCLUDE=
KUBEVIRT_INTERFACES=
ENABLE_INBOUND_IPV6=false

ip -f inet rule add fwmark 1337 lookup 133
ip -f inet route add local default dev lo table 133
Writing following contents to rules file:  /tmp/iptables-rules-1611637611221538655.txt201772342
* nat
-N ISTIO_INBOUND
-N ISTIO_REDIRECT
-N ISTIO_IN_REDIRECT
-N ISTIO_OUTPUT
-A ISTIO_INBOUND -p tcp --dport 15008 -j RETURN
-A ISTIO_REDIRECT -p tcp --dport 53 -j REDIRECT --to-ports 15053
-A ISTIO_REDIRECT -p tcp -j REDIRECT --to-ports 15001
-A ISTIO_IN_REDIRECT -p tcp -j REDIRECT --to-ports 15006
-A OUTPUT -p tcp -j ISTIO_OUTPUT
-A ISTIO_OUTPUT -o lo -s 127.0.0.6/32 -j RETURN
-A ISTIO_OUTPUT -o lo ! -d 127.0.0.1/32 -m owner --uid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -m owner --uid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -o lo ! -d 127.0.0.1/32 -m owner --gid-owner 0 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -m owner --gid-owner 0 -j RETURN
-A ISTIO_OUTPUT -o lo ! -d 127.0.0.1/32 -m owner --gid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -m owner --gid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -d 127.0.0.1/32 -j RETURN
-A ISTIO_OUTPUT -p tcp --dport 50010 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 50020 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 50075 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 50070 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 50090 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 9000 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8019 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8480 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8481 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8485 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 2181 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8030 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8031 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8032 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8033 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8088 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8040 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8041 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8042 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 10020 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 19888 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 33911 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8020 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp --dport 8082 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -d 127.1.2.3/32 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -j RETURN
-A OUTPUT -p udp --dport 53 -m owner --uid-owner 1337 -j RETURN
-A OUTPUT -p udp --dport 53 -m owner --gid-owner 0 -j RETURN
-A OUTPUT -p udp --dport 53 -m owner --gid-owner 1337 -j RETURN
-A OUTPUT -p udp --dport 53 -j DNAT --to-destination 127.0.0.1:15053
-A POSTROUTING -p udp --dport 15053 -j SNAT --to-source 127.0.0.1
COMMIT
* mangle
-N ISTIO_DIVERT
-N ISTIO_TPROXY
-N ISTIO_INBOUND
-A ISTIO_DIVERT -j MARK --set-mark 1337
-A ISTIO_DIVERT -j ACCEPT
-A ISTIO_TPROXY ! -d 127.0.0.1/32 -p tcp -j TPROXY --tproxy-mark 1337/0xffffffff --on-port 15006
-A PREROUTING -p tcp -j ISTIO_INBOUND
-A ISTIO_INBOUND -p tcp --dport 22 -j RETURN
-A ISTIO_INBOUND -p tcp --dport 15090 -j RETURN
-A ISTIO_INBOUND -p tcp --dport 15021 -j RETURN
-A ISTIO_INBOUND -p tcp --dport 8181 -j RETURN
-A ISTIO_INBOUND -p tcp --dport 15020 -j RETURN
-A ISTIO_INBOUND -p tcp -m conntrack --ctstate RELATED,ESTABLISHED -j ISTIO_DIVERT
-A ISTIO_INBOUND -p tcp -j ISTIO_TPROXY
-A PREROUTING -p tcp -m mark --mark 1337 -j CONNMARK --save-mark
-A OUTPUT -p tcp -m connmark --mark 1337 -j CONNMARK --restore-mark
-I ISTIO_INBOUND 1 -p tcp -m mark --mark 1337 -j RETURN
COMMIT

iptables-restore --noflush /tmp/iptables-rules-1611637611221538655.txt201772342
Writing following contents to rules file:  /tmp/ip6tables-rules-1611637611969044794.txt772360989

ip6tables-restore --noflush /tmp/ip6tables-rules-1611637611969044794.txt772360989
iptables-save 
# Generated by iptables-save v1.4.21 on Tue Jan 26 05:06:51 2021
*mangle
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [0:0]
:ISTIO_DIVERT - [0:0]
:ISTIO_INBOUND - [0:0]
:ISTIO_TPROXY - [0:0]
-A PREROUTING -p tcp -j ISTIO_INBOUND
-A PREROUTING -p tcp -m mark --mark 0x539 -j CONNMARK --save-mark --nfmask 0xffffffff --ctmask 0xffffffff
-A OUTPUT -p tcp -m connmark --mark 0x539 -j CONNMARK --restore-mark --nfmask 0xffffffff --ctmask 0xffffffff
-A ISTIO_DIVERT -j MARK --set-xmark 0x539/0xffffffff
-A ISTIO_DIVERT -j ACCEPT
-A ISTIO_INBOUND -p tcp -m mark --mark 0x539 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 22 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 15090 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 15021 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 8181 -j RETURN
-A ISTIO_INBOUND -p tcp -m tcp --dport 15020 -j RETURN
-A ISTIO_INBOUND -p tcp -m conntrack --ctstate RELATED,ESTABLISHED -j ISTIO_DIVERT
-A ISTIO_INBOUND -p tcp -j ISTIO_TPROXY
-A ISTIO_TPROXY ! -d 127.0.0.1/32 -p tcp -j TPROXY --on-port 15006 --on-ip 0.0.0.0 --tproxy-mark 0x539/0xffffffff
COMMIT
# Completed on Tue Jan 26 05:06:51 2021
# Generated by iptables-save v1.4.21 on Tue Jan 26 05:06:51 2021
*nat
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [0:0]
:ISTIO_INBOUND - [0:0]
:ISTIO_IN_REDIRECT - [0:0]
:ISTIO_OUTPUT - [0:0]
:ISTIO_REDIRECT - [0:0]
-A OUTPUT -p tcp -j ISTIO_OUTPUT
-A OUTPUT -p udp -m udp --dport 53 -m owner --uid-owner 1337 -j RETURN
-A OUTPUT -p udp -m udp --dport 53 -m owner --gid-owner 0 -j RETURN
-A OUTPUT -p udp -m udp --dport 53 -m owner --gid-owner 1337 -j RETURN
-A OUTPUT -p udp -m udp --dport 53 -j DNAT --to-destination 127.0.0.1:15053
-A POSTROUTING -p udp -m udp --dport 15053 -j SNAT --to-source 127.0.0.1
-A ISTIO_INBOUND -p tcp -m tcp --dport 15008 -j RETURN
-A ISTIO_IN_REDIRECT -p tcp -j REDIRECT --to-ports 15006
-A ISTIO_OUTPUT -s 127.0.0.6/32 -o lo -j RETURN
-A ISTIO_OUTPUT ! -d 127.0.0.1/32 -o lo -m owner --uid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -m owner --uid-owner 1337 -j RETURN
-A ISTIO_OUTPUT ! -d 127.0.0.1/32 -o lo -m owner --gid-owner 0 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -m owner --gid-owner 0 -j RETURN
-A ISTIO_OUTPUT ! -d 127.0.0.1/32 -o lo -m owner --gid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -m owner --gid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -d 127.0.0.1/32 -j RETURN
-A ISTIO_OUTPUT -p tcp -m tcp --dport 50010 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 50020 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 50075 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 50070 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 50090 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 9000 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8019 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8480 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8481 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8485 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 2181 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8030 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8031 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8032 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8033 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8088 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8040 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8041 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8042 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 10020 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 19888 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 33911 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8020 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -p tcp -m tcp --dport 8082 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -d 127.1.2.3/32 -j ISTIO_REDIRECT
-A ISTIO_OUTPUT -j RETURN
-A ISTIO_REDIRECT -p tcp -m tcp --dport 53 -j REDIRECT --to-ports 15053
-A ISTIO_REDIRECT -p tcp -j REDIRECT --to-ports 15001
COMMIT
# Completed on Tue Jan 26 05:06:51 2021
```

ramaraochavali commented:
So what do we do with this PR? I am still figuring out why `runAsGroup` as not being added in our environment even though injection template has it. Do we want to make `uid` and `gid` for the reasons @howardjohn mentioned or just rely on `gid` to skip?

rlenglet commented:
> Is this the capability uniquely lost in docker-shim CRI? Other CRI may support

No, I think none of them sets ambient capabilities. We can't rely on that.

rlenglet commented:
@ramaraochavali the problem seems to be specific to your environment, so I would like to first understand what's going on there before we commit anything. Thanks!

ramaraochavali commented:
@rlenglet Sure. That makes sense. I will update what I find here.

ramaraochavali commented:
I will close this PR for now and once we find what is causing `runAsGroup` not showing up and if it needs a fix, will open a separate PR.

ramaraochavali commented:
@rlenglet @lambdai Update. The "runAsGroup" not setting per container seems to be problem with EKS. Our K8s team is a following up on that.

Another question - Do we really need to be "root" (runAsNonRoot: false)  for "TPROXY" or just "NET_ADMIN" is sufficient? I think since we are running as root, the gid is automatically set to 0 and since runAsGroup is not working as expected in our env, we are stuck with gid 0. If root is really not needed, removing that also would solve our case.

rlenglet commented:
root is needed. Again, adding capabilities doesn't work in container runtimes because they don't set ambient capabilities, so capabilities are lost when changing the uid. So uid 0 is required in order to preserve `CAP_NET_ADMIN`.

rlenglet commented:
@ramaraochavali Do you have any reference about `runAsGroup` not working on EKS?

ramaraochavali commented:
@rlenglet No there is no reference any where. We could find by experiment and we did not get any response yet.

ramaraochavali commented:
@rlenglet Just closing loop on this, it turned out that there is a bug in one of the mutating webhooks that is running in our cluster is pinned to older version of client-go that does have "RunAsGroup" in the SecurityContext so was getting dropped during mutation. The webhook owner is fixing it. So It is not an AWS issue either.

rlenglet commented:
@ramaraochavali awesome, thanks for looking deeper into this!
